File: packages/runar-rb/lib/runar/compile_check.rb
# frozen_string_literal: true

# Runar compile_check — validates contracts through the Runar frontend pipeline.
#
# Runs Parse → Validate → TypeCheck (no ANF lowering or codegen) to verify
# that a contract is valid Runar that will compile to Bitcoin Script.
#
# Requires the runar_compiler gem. If not available, raises LoadError with
# installation instructions.

module Runar
  def self.compile_check(source_or_path, file_name = nil)
    if !source_or_path.include?("\n") && File.file?(source_or_path)
      source = File.read(source_or_path)
      file_name ||= source_or_path
    else
      source = source_or_path
      file_name ||= 'contract.runar.rb'
    end

    begin
      require 'runar_compiler'
    rescue LoadError
      raise LoadError,
            'compile_check requires the runar_compiler gem. ' \
            "Install it with: gem install runar_compiler, or add gem 'runar_compiler' to your Gemfile"
    end

    # Pass 1: Parse
    parse_result = RunarCompiler.send(:_parse_source, source, file_name)
    unless parse_result.errors.empty?
      raise "parse errors in #{file_name}: #{parse_result.error_strings.join('; ')}"
    end
    raise "no contract found in #{file_name}" if parse_result.contract.nil?

    # Pass 2: Validate
    val_result = RunarCompiler.send(:_validate, parse_result.contract)
    unless val_result.errors.empty?
      raise "validation errors in #{file_name}: #{val_result.error_strings.join('; ')}"
    end

    # Pass 3: Type check
    tc_result = RunarCompiler.send(:_type_check, parse_result.contract)
    unless tc_result.errors.empty?
      raise "type check errors in #{file_name}: #{tc_result.error_strings.join('; ')}"
    end

    true
  end
end

File: packages/runar-go/compile_check.go
package runar

import (
	"fmt"
	"os"
	"strings"

	"github.com/icellan/runar/compilers/go/frontend"
)

// CompileCheck runs the Rúnar frontend (parse → validate → typecheck) on a
// .runar.go contract file. Returns nil if the contract is valid Rúnar.
//
// Use this in tests alongside business logic tests to ensure the contract
// will compile to Bitcoin Script:
//
//	func TestCompile(t *testing.T) {
//	    if err := runar.CompileCheck("MyContract.runar.go"); err != nil {
//	        t.Fatalf("Rúnar compile check failed: %v", err)
//	    }
//	}
func CompileCheck(contractFile string) error {
	source, err := os.ReadFile(contractFile)
	if err != nil {
		return fmt.Errorf("reading %s: %w", contractFile, err)
	}

	result := frontend.ParseSource(source, contractFile)
	if len(result.Errors) > 0 {
		return fmt.Errorf("parse errors: %s", strings.Join(result.ErrorStrings(), "; "))
	}
	if result.Contract == nil {
		return fmt.Errorf("no contract found in %s", contractFile)
	}

	v := frontend.Validate(result.Contract)
	if len(v.Errors) > 0 {
		return fmt.Errorf("validation errors: %s", strings.Join(v.ErrorStrings(), "; "))
	}

	tc := frontend.TypeCheck(result.Contract)
	if len(tc.Errors) > 0 {
		return fmt.Errorf("type check errors: %s", strings.Join(tc.ErrorStrings(), "; "))
	}

	return nil
}
